### DISCUSSION TITLE
Are there defined cases where the context could be lost while using `ClsMiddleware`?

### DISCUSSION THREAD

**Discussion by nishkarsh-beamery** [Papooch/nestjs-cls#177] [Q&A]
> The ClsMiddleware and ClsInterceptor by default uses the safe run() method, which it should not leak context, but in some rare cases, the context can be lost.

The above statement in the [docs](https://papooch.github.io/nestjs-cls/considerations/security) mentions that while mounted as middleware, the context could be lost in rare cases. Are such cases defined anywhere?

The question is whether we can reliably use this to store information that _has to be present_ during the request lifecycle for the app to be functional. If such rare cases are not defined, then our app might completely fail without knowing the scenarios where it could fail.

**Comment by Papooch** [comment]
Please see [Compatility considerations](https://papooch.github.io/nestjs-cls/considerations/compatibility#rest) in the docs.

Context loss can happen when used with some legacy ExpressJS middlewares.

You can find some concrete examples on the internet when you search for something like "node.js express AsyncLocalStorage context lost".

If you want to find out if your application is compatible with ALS, I suggest you try it out. It either fails or doesn't. I haven't encountered any case where it _sometimes_ fails and sometimes not.
